File: src/main/java/com/tobe/healthy/schedule/repository/TrainerScheduleRepositoryCustom.java
package com.tobe.healthy.schedule.repository;

import java.time.LocalDate;
import java.util.List;
import java.util.Optional;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import com.tobe.healthy.lessonhistory.presentation.dto.in.UnwrittenLessonHistorySearchCond;
import com.tobe.healthy.schedule.presentation.dto.in.CommandRegisterSchedule;
import com.tobe.healthy.schedule.presentation.dto.out.FeedbackNotificationToTrainer;
import com.tobe.healthy.schedule.presentation.dto.out.RetrieveTrainerScheduleByLessonDtResult;
import com.tobe.healthy.schedule.domain.ReservationStatus;
import com.tobe.healthy.schedule.domain.Schedule;
import com.tobe.healthy.schedule.domain.TrainerScheduleInfo;

public interface TrainerScheduleRepositoryCustom {

	RetrieveTrainerScheduleByLessonDtResult findOneTrainerTodaySchedule(String lessonDt, Long trainerId);

	boolean validateDuplicateSchedule(
		TrainerScheduleInfo trainerScheduleInfo,
		CommandRegisterSchedule request,
		Long trainerId
	);

	Optional<Schedule> findAvailableWaitingId(Long scheduleId);

	List<Schedule> findAllSchedule(
		String lessonDt,
		LocalDate lessonStartDt,
		LocalDate lessonEndDt,
		Long trainerId
	);

	List<Schedule> findAllSchedule(
		List<Long> scheduleIds,
		List<ReservationStatus> reservationStatus,
		Long trainerId
	);

	Schedule findAllSchedule(Long scheduleId, ReservationStatus reservationStatus, Long trainerId);

	List<Schedule> findAllDisabledSchedule(LocalDate lessonStartDt, LocalDate lessonEndDt);

	List<Schedule> findAllUnwrittenLessonHistory(UnwrittenLessonHistorySearchCond request, Long memberId);

	List<Schedule> findAllSimpleLessonHistoryByMemberId(Long studentId, Long trainerId);

	Page<Schedule> findAllScheduleByStudentId(Long studentId, Pageable pageable, Long trainerId);

	List<FeedbackNotificationToTrainer> findAllFeedbackNotificationToTrainer();
}

File: src/main/kotlin/com/tobe/healthy/schedule/repository/TrainerScheduleRepositoryCustom.kt
package com.tobe.healthy.schedule.repository

import com.tobe.healthy.lessonhistory.domain.dto.`in`.UnwrittenLessonHistorySearchCond
import com.tobe.healthy.schedule.domain.dto.`in`.CommandRegisterSchedule
import com.tobe.healthy.schedule.domain.dto.`in`.RetrieveTrainerScheduleByLessonDt
import com.tobe.healthy.schedule.domain.dto.`in`.RetrieveTrainerScheduleByLessonInfo
import com.tobe.healthy.schedule.domain.dto.out.RetrieveTrainerScheduleByLessonDtResult
import com.tobe.healthy.schedule.domain.entity.ReservationStatus
import com.tobe.healthy.schedule.domain.entity.Schedule
import com.tobe.healthy.schedule.domain.entity.TrainerScheduleInfo
import java.time.LocalDate
import java.util.*

interface TrainerScheduleRepositoryCustom {
    fun findOneTrainerTodaySchedule(request: RetrieveTrainerScheduleByLessonDt, trainerId: Long): RetrieveTrainerScheduleByLessonDtResult?
    fun findOneTrainerTodaySchedule(trainerId: Long): RetrieveTrainerScheduleByLessonDtResult?
    fun validateDuplicateSchedule(trainerScheduleInfo: TrainerScheduleInfo, request: CommandRegisterSchedule, trainerId: Long): Boolean
    fun findAvailableWaitingId(scheduleId: Long): Optional<Schedule>
    fun findAllSchedule(request: RetrieveTrainerScheduleByLessonInfo, trainerId: Long): List<Schedule>
    fun findAllSchedule(scheduleIds: List<Long>, reservationStatus: ReservationStatus, trainerId: Long): List<Schedule>
    fun findAllSchedule(scheduleIds: List<Long>, reservationStatus: List<ReservationStatus>, trainerId: Long): List<Schedule>
    fun findAllSchedule(scheduleId: Long, reservationStatus: ReservationStatus, trainerId: Long): Schedule?
    fun findAllSchedule(scheduleId: Long, trainerId: Long): Schedule?
    fun findAllDisabledSchedule(lessonStartDt: LocalDate, lessonEndDt: LocalDate): List<Schedule?>
    fun findAllUnwrittenLessonHistory(request: UnwrittenLessonHistorySearchCond, memberId: Long): List<Schedule>
    fun findAllSimpleLessonHistoryByMemberId(studentId: Long, trainerId: Long): List<Schedule>
}
